### ISSUE TITLE
Update to Key Cloak 1.2.0.Final

### ISSUE THREAD

**Issue by benneq** [Smartling/spring-security-keycloak#1]
Hopefully just a simple replacement of Beta1 to Final in Gradle file.

Refs:
http://blog.keycloak.org/2015/05/keycloak-120final-released.html
http://mvnrepository.com/artifact/org.keycloak

Btw:
Will this work with Spring Security 4?

**Comment by foo4u** [comment]
Hi @benneq, we'll upgrade to 1.2.0.Final. Spring Security 3 and 4 are both supported.

Just a note on the longterm plans for this project: we contributed a true [Spring Security adapter][kc] to the Keycloak project and have a [Smartling Keycloak Extras][extras] project to fill in the gaps.The Spring Security adapter will be included in Keycloak 1.3.0 at which point we will deprecate this project and provide instructions for migrating onto the pure Spring adapter.

Look for an upgrade to 1.2.0.Final later in the week.

Best,
Scott

[kc]: https://github.com/keycloak/keycloak/tree/master/integration/spring-security
[extras]: https://github.com/Smartling/smartling-keycloak-extras

**Comment by benneq** [comment]
I just found this: https://codeoverview.wordpress.com/2015/05/15/spring-boot-rest-service-and-keycloak/
(Github Repo: https://github.com/iuliazidaru/keycloak-spring-boot-rest-angular-demo )

This works kinda okay with Spring Security 4 and Keycloak 1.2.0.

I still didn't really know, if your project is what I am looking for. From your instructions it looks like I can login / logout via a Spring Application which uses KeyCloak as it's Authentication Service.

What I was looking for: The user authenticates somehow (i.e. via Keycloak Login Page) and then the Spring Application should only use the authentication token to get the user's credentials (and roles). Will this also be possible? Using the code from above this works pretty well. For the roles I had to do some extra work.

I just wondered if there's no easier solution for this, because (as far as I understood token auth) there only must be a Spring Security filter, which detects, decrypt and validates the token.

**Comment by foo4u** [comment]
@benneq that's correct. This project - like the Spring Boot one you reference - relies on the Keycloak Tomcat Adapter to do the heavy lifting. Ultimately, what you really want is the Keycloak Spring Security adapter that we contributed to the Keycloak project. We have a fork that makes it stable until 1.3.0 is out.

I'll see if I can generate the docbook and post somewhere.

**Comment by foo4u** [comment]
Here's a preview of the Keycloak Spring Security adapter [documentation][docs].

[docs]: http://smartling.github.io/keycloak/docs/1.2.1.Smartling-SNAPSHOT/reference/en-US/html_single/#spring-security-adapter

**Comment by benneq** [comment]
Thanks :) I'll have a look at it, try it and give feedback soon.

**Comment by benneq** [comment]
First improvement: It seems like "WEB-INF/keycloak.json" is hardcoded in org.keycloak.adapters.springsecurity.AdapterDeploymentContextBean

Maybe use it as default, but provide some way to change the location. I use Spring Boot with embedded Tomcat, and there it's common to use "src/main/resources". Now I have to create "src/webapp/WEB-INF".


And the first problem: I created SecurityConfig just like in the docs. That's the configuration I expected. Pretty nice! :+1: 
```
I activated DEBUG for org.keycloak.adapters.springsecurity. When starting the application I get:
2015-05-26 08:41:15.006 DEBUG 55238 --- [ost-startStop-1] f.KeycloakAuthenticationProcessingFilter : Initializing filter 'keycloakAuthenticationProcessingFilter'
2015-05-26 08:41:15.006 DEBUG 55238 --- [ost-startStop-1] f.KeycloakAuthenticationProcessingFilter : Filter 'keycloakAuthenticationProcessingFilter' configured successfully
2015-05-26 08:41:15.006 DEBUG 55238 --- [ost-startStop-1] o.k.a.s.f.KeycloakPreAuthActionsFilter   : Initializing filter 'keycloakPreAuthActionsFilter'
2015-05-26 08:41:15.007 DEBUG 55238 --- [ost-startStop-1] o.k.a.s.f.KeycloakPreAuthActionsFilter   : Filter 'keycloakPreAuthActionsFilter' configured successfully
```

And when making a request:
```
2015-05-26 08:41:33.495 DEBUG 55238 --- [nio-8081-exec-1] f.KeycloakAuthenticationProcessingFilter : Request is to process authentication
2015-05-26 08:41:33.495 DEBUG 55238 --- [nio-8081-exec-1] f.KeycloakAuthenticationProcessingFilter : Attempting Keycloak authentication
2015-05-26 08:41:33.526  WARN 55238 --- [nio-8081-exec-1] a.s.a.SpringSecurityRequestAuthenticator : Completing bearer authentication. Bearer roles: [morethanuser, user] 
2015-05-26 08:41:33.528 DEBUG 55238 --- [nio-8081-exec-1] f.KeycloakAuthenticationProcessingFilter : Auth outcome: AUTHENTICATED
2015-05-26 08:41:33.529  INFO 55238 --- [nio-8081-exec-1] o.s.b.a.audit.listener.AuditListener     : AuditEvent [timestamp=Tue May 26 08:41:33 CEST 2015, principal=c4761e4a-3511-461e-9c45-00a77bd4dac0, type=AUTHENTICATION_SUCCESS, data={details=org.keycloak.adapters.springsecurity.account.SimpleKeycloakAccount@7fc389ea}]
2015-05-26 08:41:33.531 DEBUG 55238 --- [nio-8081-exec-1] o.k.a.s.management.HttpSessionManager    : Session created: B5FAF318657EC9A427879DB6C5C1A13E
2015-05-26 08:41:33.533 DEBUG 55238 --- [nio-8081-exec-1] f.KeycloakAuthenticationProcessingFilter : Authentication success. Updating SecurityContextHolder to contain: org.keycloak.adapters.springsecurity.token.KeycloakAuthenticationToken@4afd5149: Principal: c4761e4a-3511-461e-9c45-00a77bd4dac0; Credentials: [PROTECTED]; Authenticated: true; Details: org.keycloak.adapters.springsecurity.account.SimpleKeycloakAccount@7fc389ea; Granted Authorities: KeycloakRole{role='morethanuser'}, KeycloakRole{role='user'}
2015-05-26 08:41:33.533  INFO 55238 --- [nio-8081-exec-1] o.s.b.a.audit.listener.AuditListener     : AuditEvent [timestamp=Tue May 26 08:41:33 CEST 2015, principal=c4761e4a-3511-461e-9c45-00a77bd4dac0, type=AUTHENTICATION_SUCCESS, data={details=org.keycloak.adapters.springsecurity.account.SimpleKeycloakAccount@7fc389ea}]
```

So it seems to work, principal id and roles are correct. But my client gets a redirect response, which doesn't work. I don't know where and why it happens:
```
XMLHttpRequest cannot load http://localhost:8081/myResource. The request was redirected to 'http://localhost:8081/', which is disallowed for cross-origin requests that require preflight.
```

I tried Spring Security 4.0.1 and 3.2.7. It doesn't matter if the resource is secured or not (SecurityConfig: http.authorizeRequests().anyRequest().permitAll();). It's a default Spring Boot 1.2.3 Application, with basically zero configuration.

Do you know why this is happening? If you need more information, just tell me where I should have a look at. Then this should maybe become part of the docs.

**Comment by benneq** [comment]
I think my problem is the Redirect after successful authentication. Which happens there:

KeycloakAuthenticationProcessingFilter.doFilter(...) at the end calls successfulAuthentication(...) and this calls successHandler.onAuthenticationSuccess(...). The implementation is: SavedRequestAwareAuthenticationSuccessHandler.onAuthenticationSuccess(...) and this guy puts the Location header inside the response.

Seems like the whole Processing Filter got reworked in the next version ( https://github.com/keycloak/keycloak/blob/master/integration/spring-security/src/main/java/org/keycloak/adapters/springsecurity/filter/KeycloakAuthenticationProcessingFilter.java ). Hopefully it's fixed there. I just copied some lines of code, and don't get anyredirect anymore, but now 404.

**Comment by lukebk** [comment]
Hi @foo4u, 

I'm evaluating keycloak (1.2) for an application based on jboss and I would integrate keycloak with spring security 4. The actual adapter is not  available for jboss, correct? 



Thanks, 
Luca

**Comment by foo4u** [comment]
We haven't tested with JBoss yet, but it should work fine as long as all dependencies are in place.

**Comment by foo4u** [comment]
@benneq sorry, haven't had time to digest your feedback yet. Will hopefully get to it next week. I'm not sure why you need a CORS filter.

**Comment by benneq** [comment]
The CORS filter is needed, because I have 3 applications at different Ports:
Keycloak Server: localhost:8080
REST Backend (Spring Boot + Security): localhost:8081
AngularJS Frontend (with Spring Boot): localhost:7005

So my AJAX requests go from port 7005 to 8081. That's a CrossSite request. Spring (Boot, MVC and Security) don't provide built in CORS support. Though only requests from the same domain (and port) are accepted.
When I put the AngularJS Frontend within the port 8081 application, everything works fine.

Spring 4.2 will make it easier to configure CORS, but I think it will take still 3 months to release: http://spring.io/blog/2015/06/08/cors-support-in-spring-framework

**Comment by sajjadROR** [comment]
I'm facing following error - please guide me.
XMLHttpRequest cannot load https://keycloak-dcdevelopment.rhcloud.com/auth/realms/demo/protocol/openid-connect/token. No 'Access-Control-Allow-Origin' header is present on the requested resource. Origin 'http://localhost:8080' is therefore not allowed access.
